### PR TITLE
Fix for MongoDBHandling on ObjectToEvent map handling

### PR DESCRIPTION
**Pull request by SantoshNC68**
### Applicable Issues
NA

### Description of the Change
Made code changes to handle the updation of events in Object to Event mapper.
Also updated the logic to not store duplicate eventIds.

### Benefits
Performance improvements in storing the ObjectToEvent mapping relation without having to query multiple times.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:  @SantoshNC68 , @raja-maragani

### CODE REVIEW

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 1]
```diff
@@ -83,28 +83,29 @@ public void setJmesPathInterface(JmesPathInterface jmesPathInterface) {
...
         String eventId = getEventId(rulesObject, event);
-        String condition = "{\"_id\" : \"" + eventId + "\"}";
-        ArrayList<String> list =  getEventToObjectList(eventId);
-        boolean firstTime = list.isEmpty();
-        list = updateList(list, eventId, objectId);
-        ObjectMapper mapper = new ObjectMapper();
-        JsonNode entry = null;
-
+        String condition = "{\"_id\" : \"" + objectId + "\"}";
+        LOGGER.debug("Checking document exists in the collection with ID : {}", condition );
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:87]: For easier debugging, these two lines could be one log entry.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 2]
```diff
@@ -83,28 +83,29 @@ public void setJmesPathInterface(JmesPathInterface jmesPathInterface) {
...
+        		((ObjectNode) entry).set(listPropertyName, mapper.readTree(jsonNode.toString()));
+                final String mapStr = entry.toString(); 
+            	LOGGER.debug("MongoDbHandler Insert/Update Event: {}\nto database: {} and to Collection: {}", mapStr, databaseName, collectionName);	
+            	mongodbhandler.insertDocument(databaseName, collectionName, mapStr );
             } else {
-                mongodbhandler.updateDocument(databaseName, collectionName, condition, mapStr);
+                mongodbhandler.updateDocumentAddToSet(databaseName, collectionName, condition, eventId);
             }
         } catch (Exception e) {
             LOGGER.info("Failed to update event object list.", e);
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:105]: Perhaps info is not the best log level when an error occurs.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 3]
```diff
@@ -83,28 +83,29 @@ public void setJmesPathInterface(JmesPathInterface jmesPathInterface) {
 
     public void updateEventToObjectMapInMemoryDB(RulesObject rulesObject, String event, String objectId) {
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:84]: I would like javadoc added to this method also.

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java [hunk 4]
```diff
@@ -342,4 +343,46 @@ public void dropDatabase(String databaseName) {
         MongoDatabase db = mongoClient.getDatabase(databaseName);
         db.drop();
     }
+    
+    /**
+     * Check if the document exists
+     * @param databaseName
+     * @param collectionName
+     * @param condition is a condition for update documents
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:351]: The description of the condition parameter doesn't really make any sense. How about "a condition to find a requested object in the database"? And be consistent in the use of documents vs objects in this method and in the javadoc.

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java [hunk 5]
```diff
@@ -342,4 +343,46 @@ public void dropDatabase(String databaseName) {
...
+     * @param eventId eventId to update in the mapper collection
+     * @return 
+     */
+    public boolean updateDocumentAddToSet(String dataBaseName, String collectionName, String condition, String eventId) {
+        try {
+            MongoCollection<Document> collection = getMongoCollection(dataBaseName, collectionName);
+            if (collection != null) {
+                final Document dbObjectInput = Document.parse(condition);
+                UpdateResult updateMany = collection.updateOne(dbObjectInput, Updates.addToSet("objects", eventId));
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:378]: This method is specific to the event to object map collection, right? Perhaps clarify this in the javadoc so that it's not used elsewhere.

**Review by e-pettersson-ericsson — COMMENTED** [review]
Mostly comments on javadoc. Are there no updates needed in the tests for these changes?

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 1 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:87]: have changed it to single statement.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 2 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:105]: changed to error. Haven't modified the already existing code but have changed to error now.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 3 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:84]: Will add.

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java (hunk 5 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:378]: done

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java (hunk 4 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:351]: done

**Comment by SantoshNC68** [comment]
Fixed comments from Emelie in 74d6c4b

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 6]
```diff
@@ -81,30 +81,36 @@ public void setJmesPathInterface(JmesPathInterface jmesPathInterface) {
...
         String eventId = getEventId(rulesObject, event);
-        String condition = "{\"_id\" : \"" + eventId + "\"}";
-        ArrayList<String> list =  getEventToObjectList(eventId);
-        boolean firstTime = list.isEmpty();
-        list = updateList(list, eventId, objectId);
-        ObjectMapper mapper = new ObjectMapper();
-        JsonNode entry = null;
-
+        String condition = "{\"_id\" : \"" + objectId + "\"}";
+        LOGGER.debug("Checking document exists in the collection with ID : {}\n EventId : {}", condition, eventId);
```
Comment (raja-maragani) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:93]: ID change to condition

Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java]: changed ID to condition

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 7]
```diff
@@ -148,7 +154,7 @@ public boolean deleteEventObjectMap(String templateName) {
     }
 
     public boolean isEventInEventObjectMap(String eventId) {
-        String condition = "{\"_id\" : \"" + eventId + "\"}";
+        String condition = "{\"objects\": { \"$in\" : [\"" + eventId + "\"]} }";
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:157]: At the top of this class a variable is defined for this key called 'listPropertyName' - use it here also. However, given that you've changed the structure of this document to look like:

    ```
    "_id": "aggregated-object-id",
     "objects": ["event-id", "event-id"]
    ```
the key of the list should perhaps be called "events" instead of "objects"? It would look somehting like this instead:

    ```
    "_id": "aggregated-object-id",
     "events": ["event-id", "event-id"]
    ```

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java [hunk 8]
```diff
@@ -342,4 +343,47 @@ public void dropDatabase(String databaseName) {
...
     }
+    
+    /**
+     * Check if the document exists
+     * @param databaseName
+     * @param collectionName
+     * @param condition      a condition to find a requested object in the database
+     * @return
+     */
+    public boolean checkObjectExists(String databaseName, String collectionName, String condition) {
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:354]: I would prefer to use document in this method name, to be consistent. When we refer to objects, it should be the aggregated objects and other objects we store inside the documents in the database.

File: src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java [hunk 9]
```diff
@@ -72,8 +73,12 @@ public JsonNode runRuleOnEvent(String rule, String event) {
 
         try {
             Expression<JsonNode> expression = jmespath.compile(rule);
+            LOGGER.debug("EXPRESSION : " + expression );
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java:76]: I'm not sure how much of this logging you want to keep, but if you want to ensure they are logged together, make one entry instead of several - since the many threads in Eiffel Intelligence makes it impossible to follow otherwise.

File: src/main/java/com/ericsson/ei/jsonmerge/MergeHandler.java [hunk 10]
```diff
@@ -67,13 +67,15 @@ public String mergeObject(String id, String mergeId, RulesObject rules, String e
         try {
             // lock and get the AggregatedObject
             String aggregatedObject = getAggregatedObject(id, true);
+            LOGGER.debug("AGGREGATED OBJECT : " + aggregatedObject);
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/jsonmerge/MergeHandler.java:70]: Do you really want to log the entire aggregation twice? They can be very big, so it might clutter the log.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 11]
```diff
@@ -81,30 +81,36 @@ public void setJmesPathInterface(JmesPathInterface jmesPathInterface) {
...
     public void updateEventToObjectMapInMemoryDB(RulesObject rulesObject, String event, String objectId) {
         String eventId = getEventId(rulesObject, event);
-        String condition = "{\"_id\" : \"" + eventId + "\"}";
-        ArrayList<String> list =  getEventToObjectList(eventId);
-        boolean firstTime = list.isEmpty();
-        list = updateList(list, eventId, objectId);
-        ObjectMapper mapper = new ObjectMapper();
-        JsonNode entry = null;
-
+        String condition = "{\"_id\" : \"" + objectId + "\"}";
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:92]: With the structure of this document now changed, and the waitlist affected, some tests are probably in need of updates also.

**Review by e-pettersson-ericsson — COMMENTED** [review]
It looks like the tests still needs to be updated, but otherwise I commented mostly on logging and general naming.

**Comment by tobiasake** [comment]
What is the plan of these latest changes. Are the plan to merge this to Master ? Or just have it in this 2.0 maintenance branch and never have it on Master.
There is an internal ticket about redo the whole latest 2.x releases due to the issue of queueing up event/aggregations in Java memory instead of letting those event wait in RabbitMq instead until EI can process them instead of having this waiting in Memory which is not persistent if Queue is full or EI crash, all this event/aggregation processing is lost.

As it is now, it can't be merge directly to master. You need to implement same thing one more time, but for Master branch on PR based on Master branch code base.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 7 above)
Comment (durga-vasaadi) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java]: > At the top of this class a variable is defined for this key called 'listPropertyName' - use it here also. However, given that you've changed the structure of this document to look like:
> 
> ```
> ```
> "_id": "aggregated-object-id",
>  "objects": ["event-id", "event-id"]
> ```
> ```
> 
> the key of the list should perhaps be called "events" instead of "objects"? It would look somehting like this instead:
> 
> ```
> ```
> "_id": "aggregated-object-id",
>  "events": ["event-id", "event-id"]
> ```
> ```

changed "events" instead of "objects"

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java (hunk 8 above)
Comment (durga-vasaadi) [reply src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java]: changed the method name to "checkDocumentExists"

File: src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java (hunk 9 above)
Comment (durga-vasaadi) [reply src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java]: removed debug loggers

File: src/main/java/com/ericsson/ei/jsonmerge/MergeHandler.java (hunk 10 above)
Comment (durga-vasaadi) [reply src/main/java/com/ericsson/ei/jsonmerge/MergeHandler.java]: removed logger

**Comment by SantoshNC68** [comment]
> What is the plan of these latest changes. Are the plan to merge this to Master ? Or just have it in this 2.0 maintenance branch and never have it on Master.
> There is an internal ticket about redo the whole latest 2.x releases due to the issue of queueing up event/aggregations in Java memory instead of letting those event wait in RabbitMq instead until EI can process them instead of having this waiting in Memory which is not persistent if Queue is full or EI crash, all this event/aggregation processing is lost.
> 
> As it is now, it can't be merge directly to master. You need to implement same thing one more time, but for Master branch on PR based on Master branch code base.

Yes they will need to be updated in the master branch also and we will make it as a separate PR for master branch.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 7 above)
Comment (durga-vasaadi) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:157]: This "events" instead of "objects" changes will come in future, this changes will be release ASAP, could you please review the current changes

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java [hunk 12]
```diff
@@ -127,6 +127,7 @@ public boolean verifyAggregatedObjectExistsInDB() throws InterruptedException {
         long stopTime = System.currentTimeMillis() + 30000 + extraCheckDelay;
         while (!eventsIdList.isEmpty() && stopTime > System.currentTimeMillis()) {
             eventsIdList = compareSentEventsWithEventsInDB(eventsIdList);
+            System.out.println("******eventsIdList********"+eventsIdList.size());
```
Comment (e-pettersson-ericsson) [inline src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java:130]: Remove these when the tests are working again.

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java [hunk 13]
```diff
@@ -145,12 +146,14 @@ public boolean verifyAggregatedObjectExistsInDB() throws InterruptedException {
     private List<String> compareSentEventsWithEventsInDB(List<String> checklist) {
         mongoClient = new MongoClient(getMongoDbHost(), getMongoDbPort());
         MongoDatabase db = mongoClient.getDatabase(database);
+        System.out.println("*******checklist*******"+checklist);
```
Comment (e-pettersson-ericsson) [inline src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java:149]: Same with this one, remove this.

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java [hunk 14]
```diff
@@ -145,12 +146,14 @@ public boolean verifyAggregatedObjectExistsInDB() throws InterruptedException {
...
         MongoCollection<Document> collection = db.getCollection(eventMapCollection);
         List<Document> documents = collection.find().into(new ArrayList<>());
         for (Document document : documents) {
             for (String expectedID : new ArrayList<>(checklist)) {
-                if (expectedID.equals(document.get("_id").toString())) {
-                    checklist.remove(expectedID);
+            	Document objects = (Document) document.get("objects"))
```
Comment (e-pettersson-ericsson) [inline src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java:154]: Shouldn't this be "events"?

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java [hunk 15]
```diff
@@ -81,30 +81,36 @@ public void setJmesPathInterface(JmesPathInterface jmesPathInterface) {
...
-                mongodbhandler.insertDocument(databaseName, collectionName, mapStr);
+        	if (!docExists) {
+        		ArrayList<String> list =  new ArrayList<String>();
+        		list.add(eventId);
+        		final ObjectMapper mapper = new ObjectMapper();
+    			JsonNode entry = new ObjectMapper().readValue(condition, JsonNode.class);
+        		ArrayNode jsonNode = mapper.convertValue(list, ArrayNode.class);
+        		((ObjectNode) entry).set(listPropertyName, mapper.readTree(jsonNode.toString()));
+                final String mapStr = entry.toString(); 
+            	LOGGER.debug("MongoDbHandler Insert/Update Event: {}\nto database: {} and to Collection: {}", mapStr, databaseName, collectionName);
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:104]: I wonder if this is supposed to be logged on INFO level? But that can be handled later, if needed.

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java [hunk 16]
```diff
@@ -342,4 +343,47 @@ public void dropDatabase(String databaseName) {
...
+        Document doc = mongoCollection.find(BasicDBObject.parse(condition)).first();
+        if (doc == null || doc.isEmpty()) {
+            return false;
+        }
+        return true;
+    }
+    
+    
+    /**
+     * Update the existing doucment with unique objects list
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:366]: doucment --> document

File: src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java [hunk 17]
```diff
@@ -72,8 +72,10 @@ public JsonNode runRuleOnEvent(String rule, String event) {
 
         try {
             Expression<JsonNode> expression = jmespath.compile(rule);
+            LOGGER.debug("EXPRESSION : {}", expression );
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java:75]: These 2 log statements could be merged into one, so it's easier to follow.

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java (hunk 14 above)
Comment (e-pettersson-ericsson) [reply src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java:154]: Travis complains a lot about compilation errors, which can be discovered earlier if you run the maven commands in your local environment.

File: src/main/java/com/ericsson/ei/waitlist/WaitListStorageHandler.java [hunk 18]
```diff
@@ -82,8 +82,8 @@ public void addEventToWaitListIfNotExisting(String event, RulesObject rulesObjec
         }
     }
 
-    private String findEventInWaitList(JsonNode id) {
-        String condition = "{\"_id\" : \"" + id + "\"}";
+    private String findEventInWaitList(String id) {
+    	String condition = "{\"_id\" : \"" + id + "\"}";
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/waitlist/WaitListStorageHandler.java:86]: Why the change to 2 spaces here? Keep it the formatting the same in the entire file

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 11 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:92]: No it will not be affected as the query mechanism from the waitlist handler has been changed to search instead in the events (previously objects)

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java (hunk 12 above)
Comment (SantoshNC68) [reply src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java]: removed.

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java (hunk 13 above)
Comment (SantoshNC68) [reply src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java]: removed

File: src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java (hunk 14 above)
Comment (SantoshNC68) [reply src/functionaltests/java/com/ericsson/ei/utils/DataBaseManager.java]: It should be events. Just wanted to check why the travis has been failing.

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 15 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:104]: yes we believe this can be kept as debug as long as we don't need them to be on the information level.

File: src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java (hunk 17 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/jmespath/JmesPathInterface.java:75]: ok

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 7 above)
Comment (e-pettersson-ericsson) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:157]: I see another PR open, which duplicates this one. Please update the changes in this PR and close the other one (then we keep the review history also) It's possible for @SantoshNC68  to allow @durga-vasaadi as collaborator on your fork, and this way you can both push to this one PR.

File: src/main/java/com/ericsson/ei/waitlist/WaitListStorageHandler.java (hunk 18 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/waitlist/WaitListStorageHandler.java:86]: We have removed tabs, may be because of that it looks like this. will update.

File: src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java (hunk 16 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/MongoDBHandler.java:366]: done

File: src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java (hunk 7 above)
Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:157]: Yes I have done that and we have worked together, but because of the issues on the travis we have another  PR. We will close that.

Comment (SantoshNC68) [reply src/main/java/com/ericsson/ei/handlers/EventToObjectMapHandler.java:157]: Also we will not make changes for events in this PR but will do it another PR as we need to see the impact of that.

**Review by e-pettersson-ericsson — APPROVED** [review]
I haven't been able to test this locally due to my Java environment, but since I know it's been tested in a real environment and I realize time is of the essence for this fix, I'll approve it. But I hope the remaining comments will be fixed in an upcoming PR as soon as possible. And that these changes are also made on master if the same issue occurs there.
